o1vm/src/pickles/prover.rs

```diff
@@ -0,0 +1,221 @@
+        };
+        // Doing in parallel
+        let scratch = scratch.into_par_iter().map(eval_col).collect::<Vec<_>>();
+        WitnessColumns {
+            scratch: scratch.try_into().unwrap(),
+            instruction_counter: eval_col(instruction_counter),
+            error: eval_col(error.clone()),
+            // FIXME
+            selector: eval_col(error),
+        }
+    };
+    let commitments = {
+        let WitnessColumns {
+            scratch,
+            instruction_counter,
+            error,
+            // FIXME
+            selector: _,
+        } = &polys;
+        // Note: we do not blind. We might want in the near future in case we
+        // have a column with only zeroes.
+        let comm = |poly: &DensePolynomial<G::ScalarField>| srs.commit_non_hiding(poly, num_chunks);
+        // Doing in parallel
+        let scratch = scratch.par_iter().map(comm).collect::<Vec<_>>();
+        WitnessColumns {
+            scratch: scratch.try_into().unwrap(),
+            instruction_counter: comm(instruction_counter),
+            error: comm(error),
+            // FIXME
+            selector: comm(error),
+        }
+    };
+
+    // Absorbing the commitments - Fiat Shamir
+    // We do not parallelize as we need something deterministic.
+    for comm in commitments.scratch.iter() {
+        absorb_commitment(&mut fq_sponge, comm)
+    }
+    absorb_commitment(&mut fq_sponge, &commitments.instruction_counter);
+    absorb_commitment(&mut fq_sponge, &commitments.error);
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Round 2: Creating and committing to the quotient polynomial
+    ////////////////////////////////////////////////////////////////////////////
+
+    // FIXME: add quotient polynomial
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Round 3: Evaluations at ζ and ζω
+    ////////////////////////////////////////////////////////////////////////////
+
+    let zeta_chal = ScalarChallenge(fq_sponge.challenge());
+    let (_, endo_r) = G::endos();
+
+    let zeta = zeta_chal.to_field(endo_r);
+    let zeta_omega = zeta * omega;
+
+    // FIXME: add selectors
+    let evals = |point| {
+        let WitnessColumns {
+            scratch,
+            instruction_counter,
+            error,
+            // FIXME
+            selector: _,
+        } = &polys;
+        let eval = |poly: &DensePolynomial<G::ScalarField>| poly.evaluate(point);
+        let scratch = scratch.par_iter().map(eval).collect::<Vec<_>>();
+        WitnessColumns {
+            scratch: scratch.try_into().unwrap(),
+            instruction_counter: eval(instruction_counter),
+            error: eval(error),
+            // FIXME
+            selector: eval(error),
+        }
+    };
+    // All evaluations at ζ
+    let zeta_evaluations = evals(&zeta);
+    // All evaluations at ζω
+    let zeta_omega_evaluations = evals(&zeta_omega);
```

Comment:
it's unclear to me which poly should be evaluated at zeta_omega
Also this info can (should?) be derived from the constraints rather than hardcoded

Reply:
> Also this info can (should?) be derived from the constraints rather than hardcoded

Yes, we will probably use it with the lookup argument soon.